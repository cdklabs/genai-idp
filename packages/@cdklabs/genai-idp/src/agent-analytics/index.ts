/*
Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
SPDX-License-Identifier: Apache-2.0
*/

export * from "./agent-analytics";
export * from "./functions";
