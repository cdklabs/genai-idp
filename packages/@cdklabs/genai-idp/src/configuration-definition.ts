/*
Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
SPDX-License-Identifier: Apache-2.0
*/

/**
 * Transforms a value in a nested object based on a flat path.
 * Used internally to apply transformations to specific configuration properties.
 *
 * @param obj The object to transform
 * @param flatPath Array of keys representing the path to the target property
 * @param transform Function to transform the target value
 * @private
 */
function transformValue(
  obj: any,
  flatPath: string[],
  transform: (value: any) => any,
) {
  if (!obj || typeof obj !== "object" || flatPath.length === 0) {
    return;
  }

  const [currentKey, ...remainingPath] = flatPath;

  if (remainingPath.length === 0) {
    // We've reached the target path, set the value
    obj[currentKey] = transform(obj[currentKey]);
  } else if (currentKey in obj && typeof obj[currentKey] === "object") {
    // Continue traversing the path
    transformValue(obj[currentKey], remainingPath, transform);
  }
}

/**
 * Interface for configuration definitions.
 * Provides methods to access configuration data.
 */
export interface IConfigurationDefinition {
  /**
   * Gets the raw configuration object.
   *
   * @returns The configuration as a JavaScript object
   */
  raw(): { [key: string]: any };
}

/**
 * Defines a transformation to apply to a specific property in the configuration.
 * Used to modify configuration values during initialization.
 */
export interface IConfigurationDefinitionPropertyTransform {
  /**
   * Dot-notation path to the property to transform (e.g., "extraction.model").
   */
  readonly flatPath: string;

  /**
   * Function to transform the property value.
   *
   * @param value The original property value
   * @returns The transformed property value
   */
  transform(value: any): any;
}

/**
 * Properties for creating a configuration definition.
 */
export interface ConfigurationDefinitionProps {
  /**
   * The configuration object to use.
   * Contains all settings for the document processing pipeline.
   */
  readonly configurationObject: { [key: string]: any };

  /**
   * Optional transformations to apply to specific properties.
   * Used to modify configuration values during initialization.
   */
  readonly transforms?: IConfigurationDefinitionPropertyTransform[];
}

/**
 * A configuration definition for document processing.
 * Manages configuration data and provides methods to access it.
 */
export class ConfigurationDefinition implements IConfigurationDefinition {
  /**
   * The processed configuration contents.
   * @private
   */
  private readonly contents: { [key: string]: any };

  /**
   * Creates a new ConfigurationDefinition.
   *
   * @param props Properties for the configuration definition
   */
  constructor(props: ConfigurationDefinitionProps) {
    const raw = JSON.parse(JSON.stringify(props.configurationObject));
    props.transforms
      ?.filter((override) => !!override.transform)
      .forEach(({ flatPath: path, transform }) => {
        transformValue(raw, path.split("."), transform);
      });
    this.contents = raw;
  }

  /**
   * Gets the raw configuration object.
   *
   * @returns The configuration as a JavaScript object
   */
  raw() {
    return this.contents;
  }
}
