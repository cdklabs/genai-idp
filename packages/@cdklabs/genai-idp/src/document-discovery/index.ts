/*
Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
SPDX-License-Identifier: Apache-2.0
*/

export * from "./discovery-table";
export * from "./discovery-queue";
export * from "./document-discovery";
export * from "./functions";
