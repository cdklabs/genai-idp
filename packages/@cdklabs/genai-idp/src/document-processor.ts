/*
Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
SPDX-License-Identifier: Apache-2.0
*/

import * as sfn from "aws-cdk-lib/aws-stepfunctions";
import { IConstruct } from "constructs";
import { IProcessingEnvironment } from "./processing-environment";

/**
 * Properties required to configure a document processor implementation.
 * Document processors are responsible for extracting structured data from unstructured documents
 * using various AI/ML services and processing patterns.
 *
 * The GenAI IDP Accelerator provides multiple processor implementations to handle different
 * document processing scenarios, from standard forms to complex specialized documents.
 */
export interface DocumentProcessorProps {
  /**
   * The maximum number of documents that can be processed concurrently.
   * Controls the throughput and resource utilization of the document processing system.
   *
   * @default 100 concurrent workflows
   */
  readonly maxProcessingConcurrency?: number;

  /**
   * The processing environment that provides shared infrastructure and services.
   * Contains input/output buckets, tracking tables, API endpoints, and other
   * resources needed for document processing operations.
   */
  readonly environment: IProcessingEnvironment;
}

/**
 * Interface for document processor implementations.
 * Document processors handle the extraction of structured data from documents
 * using different processing patterns and AI/ML services.
 *
 * The GenAI IDP Accelerator includes multiple processor implementations:
 * - Pattern 1: Uses Amazon Bedrock Data Automation for document processing with minimal custom code
 * - Pattern 2: Implements custom extraction using Amazon Bedrock foundation models for flexible processing
 * - Pattern 3: Provides specialized document processing using SageMaker endpoints for custom classification models
 *
 * Each pattern is optimized for different document types, complexity levels, and customization needs.
 */
export interface IDocumentProcessor extends IConstruct {
  /**
   * The Step Functions state machine that orchestrates the document processing workflow.
   * Manages the sequence of processing steps and handles error conditions.
   * This state machine is triggered for each document that needs processing
   * and coordinates the entire extraction pipeline.
   */
  readonly stateMachine: sfn.IStateMachine;

  /**
   * The maximum number of documents that can be processed concurrently.
   * Controls the throughput and resource utilization of the document processing system.
   */
  readonly maxProcessingConcurrency: number;

  /**
   * The processing environment that provides shared infrastructure and services.
   * Contains input/output buckets, tracking tables, API endpoints, and other
   * resources needed for document processing operations.
   */
  readonly environment: IProcessingEnvironment;
}
