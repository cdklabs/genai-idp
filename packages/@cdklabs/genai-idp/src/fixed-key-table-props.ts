// ~~ Generated by projen. To modify, edit .projenrc.js and run "npx projen".
import type { aws_dynamodb, aws_iam, aws_kinesis, aws_kms, Duration, RemovalPolicy } from 'aws-cdk-lib';

/**
 * Properties for a DynamoDB Table that has a predefined, fixed partitionKey, sortKey, and timeToLiveAttribute
 */
export interface FixedKeyTableProps {
  /**
   * Enforces a particular physical table name.
   * @default <generated>
   * @stability stable
   */
  readonly tableName?: string;
  /**
   * Kinesis Data Stream to capture item-level changes for the table.
   * @default - no Kinesis Data Stream
   * @stability stable
   */
  readonly kinesisStream?: aws_kinesis.IStream;
  /**
   * Kinesis Data Stream approximate creation timestamp precision.
   * @default ApproximateCreationDateTimePrecision.MICROSECOND
   * @stability stable
   */
  readonly kinesisPrecisionTimestamp?: aws_dynamodb.ApproximateCreationDateTimePrecision;
  /**
   * The write capacity for the table.
   * Careful if you add Global Secondary Indexes, as
   * those will share the table's provisioned throughput.
   *
   * Can only be provided if billingMode is Provisioned.
   * @default 5
   * @stability stable
   */
  readonly writeCapacity?: number;
  /**
   * Specify values to pre-warm you DynamoDB Table Warm Throughput feature is not available for Global Table replicas using the `Table` construct.
   * To enable Warm Throughput, use the `TableV2` construct instead.
   * @default - warm throughput is not configured
   * @stability stable
   */
  readonly warmThroughput?: aws_dynamodb.WarmThroughput;
  /**
   * [WARNING: Use this flag with caution, misusing this flag may cause deleting existing replicas, refer to the detailed documentation for more information] Indicates whether CloudFormation stack waits for replication to finish.
   * If set to false, the CloudFormation resource will mark the resource as
   * created and replication will be completed asynchronously. This property is
   * ignored if replicationRegions property is not set.
   *
   * WARNING:
   * DO NOT UNSET this property if adding/removing multiple replicationRegions
   * in one deployment, as CloudFormation only supports one region replication
   * at a time. CDK overcomes this limitation by waiting for replication to
   * finish before starting new replicationRegion.
   *
   * If the custom resource which handles replication has a physical resource
   * ID with the format `region` instead of `tablename-region` (this would happen
   * if the custom resource hasn't received an event since v1.91.0), DO NOT SET
   * this property to false without making a change to the table name.
   * This will cause the existing replicas to be deleted.
   * @default true
   * @stability stable
   */
  readonly waitForReplicationToFinish?: boolean;
  /**
   * Specify the table class.
   * @default STANDARD
   * @stability stable
   */
  readonly tableClass?: aws_dynamodb.TableClass;
  /**
   * When an item in the table is modified, StreamViewType determines what information is written to the stream for this table.
   * @default - streams are disabled unless `replicationRegions` is specified
   * @stability stable
   */
  readonly stream?: aws_dynamodb.StreamViewType;
  /**
   * Resource policy to assign to table.
   * @default - No resource policy statement
   * @stability stable
   */
  readonly resourcePolicy?: aws_iam.PolicyDocument;
  /**
   * The timeout for a table replication operation in a single region.
   * @default Duration.minutes(30)
   * @stability stable
   */
  readonly replicationTimeout?: Duration;
  /**
   * Regions where replica tables will be created.
   * @default - no replica tables are created
   * @stability stable
   */
  readonly replicationRegions?: Array<string>;
  /**
   * The removal policy to apply to the DynamoDB replica tables.
   * @default undefined - use DynamoDB Table's removal policy
   * @stability stable
   */
  readonly replicaRemovalPolicy?: RemovalPolicy;
  /**
   * The removal policy to apply to the DynamoDB Table.
   * @default RemovalPolicy.RETAIN
   * @stability stable
   */
  readonly removalPolicy?: RemovalPolicy;
  /**
   * The read capacity for the table.
   * Careful if you add Global Secondary Indexes, as
   * those will share the table's provisioned throughput.
   *
   * Can only be provided if billingMode is Provisioned.
   * @default 5
   * @stability stable
   */
  readonly readCapacity?: number;
  /**
   * Whether point-in-time recovery is enabled and recoveryPeriodInDays is set.
   * @default - point in time recovery is not enabled.
   * @stability stable
   */
  readonly pointInTimeRecoverySpecification?: aws_dynamodb.PointInTimeRecoverySpecification;
  /**
   * Whether point-in-time recovery is enabled.
   * @default false - point in time recovery is not enabled.
   * @deprecated use `pointInTimeRecoverySpecification` instead
   * @stability deprecated
   */
  readonly pointInTimeRecovery?: boolean;
  /**
   * The write request units for the table.
   * Careful if you add Global Secondary Indexes, as
   * those will share the table's maximum on-demand throughput.
   *
   * Can only be provided if billingMode is PAY_PER_REQUEST.
   * @default - on-demand throughput is disabled
   * @stability stable
   */
  readonly maxWriteRequestUnits?: number;
  /**
   * The maximum read request units for the table.
   * Careful if you add Global Secondary Indexes, as
   * those will share the table's maximum on-demand throughput.
   *
   * Can only be provided if billingMode is PAY_PER_REQUEST.
   * @default - on-demand throughput is disabled
   * @stability stable
   */
  readonly maxReadRequestUnits?: number;
  /**
   * The properties of data being imported from the S3 bucket source to the table.
   * @default - no data import from the S3 bucket
   * @stability stable
   */
  readonly importSource?: aws_dynamodb.ImportSourceSpecification;
  /**
   * External KMS key to use for table encryption.
   * This property can only be set if `encryption` is set to `TableEncryption.CUSTOMER_MANAGED`.
   * @default - If `encryption` is set to `TableEncryption.CUSTOMER_MANAGED` and this
property is undefined, a new KMS key will be created and associated with this table.
If `encryption` and this property are both undefined, then the table is encrypted with
an encryption key managed by DynamoDB, and you are not charged any fee for using it.
   * @stability stable
   */
  readonly encryptionKey?: aws_kms.IKey;
  /**
   * Whether server-side encryption with an AWS managed customer master key is enabled.
   * This property cannot be set if `serverSideEncryption` is set.
   *
   * > **NOTE**: if you set this to `CUSTOMER_MANAGED` and `encryptionKey` is not
   * > specified, the key that the Tablet generates for you will be created with
   * > default permissions. If you are using CDKv2, these permissions will be
   * > sufficient to enable the key for use with DynamoDB tables.  If you are
   * > using CDKv1, make sure the feature flag
   * > `@aws-cdk/aws-kms:defaultKeyPolicies` is set to `true` in your `cdk.json`.
   * @default - The table is encrypted with an encryption key managed by DynamoDB, and you are not charged any fee for using it.
   * @stability stable
   */
  readonly encryption?: aws_dynamodb.TableEncryption;
  /**
   * Enables deletion protection for the table.
   * @default false
   * @stability stable
   */
  readonly deletionProtection?: boolean;
  /**
   * Whether CloudWatch contributor insights is enabled and what mode is selected.
   * @default - contributor insights is not enabled
   * @stability stable
   */
  readonly contributorInsightsSpecification?: aws_dynamodb.ContributorInsightsSpecification;
  /**
   * Whether CloudWatch contributor insights is enabled.
   * @default false
   * @deprecated use `contributorInsightsSpecification instead
   * @stability deprecated
   */
  readonly contributorInsightsEnabled?: boolean;
  /**
   * Specify how you are charged for read and write throughput and how you manage capacity.
   * @default PROVISIONED if `replicationRegions` is not specified, PAY_PER_REQUEST otherwise
   * @stability stable
   */
  readonly billingMode?: aws_dynamodb.BillingMode;
}
