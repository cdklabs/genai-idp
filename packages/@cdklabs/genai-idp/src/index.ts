/*
Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
SPDX-License-Identifier: Apache-2.0
*/

export * from "./functions";
export * from "./hitl";
export * from "./reporting";
export * from "./vpc-configuration";
export * from "./fixed-key-table-props";
export * from "./concurrency-table";
export * from "./configuration-definition";
export * from "./configuration-table";
export * from "./tracking-table";
export * from "./processing-environment-api";
export * from "./document-processor";
export * from "./processing-environment";
export * from "./user-identity";
export * from "./web-application";
export * from "./idp-python-layer-version";
export * from "./log-level";
export * from "./custom-prompt-generator";
export * from "./agent-analytics";
export * from "./document-discovery";
export * from "./internal";
