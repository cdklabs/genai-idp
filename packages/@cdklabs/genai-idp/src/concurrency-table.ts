/*
Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
SPDX-License-Identifier: Apache-2.0
*/

import * as path from "path";
import * as lambda_python from "@aws-cdk/aws-lambda-python-alpha";
import * as cdk from "aws-cdk-lib";
import { AttributeType, ITable, Table } from "aws-cdk-lib/aws-dynamodb";
import * as lambda from "aws-cdk-lib/aws-lambda";

import { Construct } from "constructs";
import { FixedKeyTableProps as ConcurrencyTableProps } from "./fixed-key-table-props";

/**
 * Interface for the concurrency management table.
 * This table is used to track and limit concurrent document processing tasks,
 * preventing resource exhaustion and ensuring system stability under load.
 */
export interface IConcurrencyTable extends ITable {}

/**
 * A DynamoDB table for managing concurrency limits in document processing.
 *
 * This construct creates a table with a custom resource that initializes
 * concurrency counters, allowing the system to control how many documents
 * are processed simultaneously to prevent resource exhaustion.
 */
export class ConcurrencyTable extends Table implements IConcurrencyTable {
  /**
   * Creates a new ConcurrencyTable.
   *
   * @param scope The construct scope
   * @param id The construct ID
   * @param props Configuration properties for the DynamoDB table
   */
  constructor(scope: Construct, id: string, props?: ConcurrencyTableProps) {
    super(scope, id, {
      ...props,
      partitionKey: { name: "counter_id", type: AttributeType.STRING },
    });

    // Lambda Function to initialize the concurrency table
    const initializeConcurrencyTableLambda = new lambda_python.PythonFunction(
      this,
      "InitializeConcurrencyTableLambda",
      {
        runtime: lambda.Runtime.PYTHON_3_12,
        entry: path.join(
          __dirname,
          "..",
          "assets",
          "lambdas",
          "initialize_counter",
        ),
        bundling: {
          commandHooks: {
            beforeBundling: (_i: string, _o: string): string[] => {
              return [];
            },
            afterBundling: (_i: string, outputDir: string): string[] => {
              return [
                `find ${outputDir} -type d -name "*.dist-info" -exec rm -rf {} +`,
                `find ${outputDir} -type d -name "*.egg-info" -exec rm -rf {} +`,
                `find ${outputDir} -type d -name "__pycache__" -exec rm -rf {} +`,
                `find ${outputDir} -type d -name "build" -exec rm -rf {} +`,
                `find ${outputDir} -type d -name "tests" -exec rm -rf {} +`,
              ];
            },
          },
        },
        timeout: cdk.Duration.seconds(30),
        environment: {
          CONCURRENCY_TABLE: this.tableName,
        },
      },
    );

    // Custom resource to initialize the concurrency table
    new cdk.CustomResource(this, "InitializeConcurrencyTableCustomResource", {
      serviceToken: initializeConcurrencyTableLambda.functionArn,
      properties: {
        TableName: this.tableName,
      },
    });

    this.grantReadWriteData(initializeConcurrencyTableLambda);
  }
}
