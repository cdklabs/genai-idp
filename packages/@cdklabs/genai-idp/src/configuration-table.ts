/*
Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
SPDX-License-Identifier: Apache-2.0
*/

import { AttributeType, ITable, Table } from "aws-cdk-lib/aws-dynamodb";

import { Construct } from "constructs";
import { FixedKeyTableProps as ConfigurationTableProps } from "./fixed-key-table-props";

/**
 * Interface for the configuration management table.
 * This table stores system-wide configuration settings for the document processing solution,
 * including extraction schemas, model parameters, evaluation criteria, and UI settings.
 */
export interface IConfigurationTable extends ITable {}

/**
 * A DynamoDB table for storing configuration settings for the document processing solution.
 *
 * This table uses a fixed partition key "Configuration" to store various configuration
 * items such as extraction schemas, evaluation settings, and system parameters.
 * It provides a centralized location for managing configuration that can be
 * accessed by multiple components of the solution.
 *
 * Configuration items stored in this table can include:
 * - Document extraction schemas and templates
 * - Model parameters and prompt configurations
 * - Evaluation criteria and thresholds
 * - UI settings and customizations
 * - Processing workflow configurations
 */
export class ConfigurationTable extends Table implements IConfigurationTable {
  /**
   * Creates a new ConfigurationTable.
   *
   * @param scope The construct scope
   * @param id The construct ID
   * @param props Configuration properties for the DynamoDB table
   */
  constructor(scope: Construct, id: string, props?: ConfigurationTableProps) {
    super(scope, id, {
      ...props,
      partitionKey: { name: "Configuration", type: AttributeType.STRING },
    });
  }
}
