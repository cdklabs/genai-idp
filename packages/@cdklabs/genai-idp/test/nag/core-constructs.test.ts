/*
Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
SPDX-License-Identifier: Apache-2.0
*/

import * as cxapi from "@aws-cdk/cx-api";
import { App, Aspects, Stack } from "aws-cdk-lib";
import { Annotations, Match, Template } from "aws-cdk-lib/assertions";
import { AwsSolutionsChecks, NagSuppressions } from "cdk-nag";
import { IdpPythonLayerVersion } from "../../src";

describe("Core Constructs CDK Nag Compliance", () => {
  let app: App;
  let stack: Stack;

  beforeEach(() => {
    // Create app with bundling disabled for unit tests
    app = new App({
      context: {
        [cxapi.BUNDLING_STACKS]: [], // Use correct context key to disable bundling
        "@aws-cdk/aws-lambda:recognizeLayerVersion": true,
        "@aws-cdk/aws-lambda:recognizeVersionProps": true,
      },
    });
    stack = new Stack(app, "TestStack");
    Aspects.of(stack).add(new AwsSolutionsChecks());
  });

  describe("IdpPythonLayerVersion", () => {
    test("satisfies CDK NAG AwsSolutionsChecks with minimal configuration", () => {
      const layer = IdpPythonLayerVersion.getOrCreate(stack, "core");

      // Verify bundling is disabled before template synthesis
      expect(stack.bundlingRequired).toBe(false);

      // Apply any necessary suppressions for known acceptable violations
      NagSuppressions.addResourceSuppressionsByPath(
        stack,
        `/${layer.node.path}`,
        [
          {
            id: "AwsSolutions-L1",
            reason:
              "Lambda layer uses supported Python runtime version as required by the IDP framework",
          },
        ],
      );

      // Trigger CDK Nag analysis
      app.synth();

      const template = Template.fromStack(stack);
      const warnings = Annotations.fromStack(stack).findWarning(
        "*",
        Match.stringLikeRegexp("AwsSolutions-.*"),
      );
      const errors = Annotations.fromStack(stack).findError(
        "*",
        Match.stringLikeRegexp("AwsSolutions-.*"),
      );

      expect(warnings).toHaveLength(0);
      expect(errors).toHaveLength(0);

      expect(template.toJSON()).toMatchSnapshot();
    });

    test("satisfies CDK NAG with comprehensive module configuration", () => {
      const layer = IdpPythonLayerVersion.getOrCreate(
        stack,
        "ocr",
        "classification",
        "assessment",
      );

      // Verify bundling is disabled before template synthesis
      expect(stack.bundlingRequired).toBe(false);

      // Apply suppressions for acceptable violations
      NagSuppressions.addResourceSuppressionsByPath(
        stack,
        `/${layer.node.path}`,
        [
          {
            id: "AwsSolutions-L1",
            reason:
              "Lambda layer uses supported Python runtime version as required by the IDP framework with multiple modules",
          },
        ],
      );

      // Trigger CDK Nag analysis
      app.synth();

      const template = Template.fromStack(stack);
      const warnings = Annotations.fromStack(stack).findWarning(
        "*",
        Match.stringLikeRegexp("AwsSolutions-.*"),
      );
      const errors = Annotations.fromStack(stack).findError(
        "*",
        Match.stringLikeRegexp("AwsSolutions-.*"),
      );

      expect(warnings).toHaveLength(0);
      expect(errors).toHaveLength(0);

      expect(template.toJSON()).toMatchSnapshot();
    });

    test("detects CDK Nag violations when suppressions are not applied", () => {
      IdpPythonLayerVersion.getOrCreate(stack, "core");

      // Verify bundling is disabled before template synthesis
      expect(stack.bundlingRequired).toBe(false);

      // Do not apply suppressions to test violation detection
      // Trigger CDK Nag analysis
      app.synth();

      const warnings = Annotations.fromStack(stack).findWarning(
        "*",
        Match.stringLikeRegexp("AwsSolutions-.*"),
      );
      const errors = Annotations.fromStack(stack).findError(
        "*",
        Match.stringLikeRegexp("AwsSolutions-.*"),
      );

      const template = Template.fromStack(stack);
      const resources = template.findResources("*");
      const resourceCount = Object.keys(resources).length;

      // If no resources are created (due to bundling being disabled),
      // then no violations should be detected
      if (resourceCount === 0) {
        expect(warnings).toHaveLength(0);
        expect(errors).toHaveLength(0);
        console.log(
          "No resources created due to bundling being disabled - no violations expected",
        );
      } else {
        // If resources are created, violations should be detected when suppressions are not applied
        const totalViolations = warnings.length + errors.length;
        expect(totalViolations).toBeGreaterThan(0);

        // Log violations for debugging
        if (warnings.length > 0) {
          console.log(
            "CDK Nag warnings detected:",
            warnings.map((w) => w.entry.data),
          );
        }
        if (errors.length > 0) {
          console.log(
            "CDK Nag errors detected:",
            errors.map((e) => e.entry.data),
          );
        }
      }
    });
  });

  describe("Security Best Practices", () => {
    test("layer version follows security best practices", () => {
      // Verify bundling is disabled before calling getOrCreate
      expect(stack.bundlingRequired).toBe(false);

      const layer = IdpPythonLayerVersion.getOrCreate(stack, "core");

      // Verify the layer construct is created
      expect(layer).toBeDefined();
      expect(layer.layerVersionArn).toBeDefined();

      const template = Template.fromStack(stack);

      // Check if layer resource exists, if bundling is disabled it might not synthesize
      const layerResources = template.findResources(
        "AWS::Lambda::LayerVersion",
      );
      if (Object.keys(layerResources).length > 0) {
        // Verify layer has proper configuration
        template.hasResourceProperties("AWS::Lambda::LayerVersion", {
          CompatibleRuntimes: Match.anyValue(),
          Content: {
            S3Bucket: Match.anyValue(),
            S3Key: Match.anyValue(),
          },
        });

        // Ensure no hardcoded sensitive values in environment variables or properties
        const templateJson = template.toJSON();
        const templateString = JSON.stringify(templateJson);

        // Check for hardcoded passwords, secrets, or API keys (but not S3 keys)
        expect(templateString).not.toMatch(/password\s*[:=]/i);
        expect(templateString).not.toMatch(/secret\s*[:=]/i);
        expect(templateString).not.toMatch(/api[_-]?key\s*[:=]/i);
        expect(templateString).not.toMatch(/access[_-]?key\s*[:=]/i);
      }
    });
  });

  describe("Performance and Cost Optimization", () => {
    test("layer configuration supports performance optimization", () => {
      // Verify bundling is disabled before calling getOrCreate
      expect(stack.bundlingRequired).toBe(false);

      // Test that minimal layer configuration is supported for performance
      const minimalLayer = IdpPythonLayerVersion.getOrCreate(stack, "core");

      // Test that comprehensive layer configuration is available when needed
      const comprehensiveLayer = IdpPythonLayerVersion.getOrCreate(
        stack,
        "ocr",
        "classification",
        "assessment",
        "extraction",
      );

      // Verify both layers are created at construct level
      expect(minimalLayer).toBeDefined();
      expect(comprehensiveLayer).toBeDefined();
      expect(minimalLayer).not.toBe(comprehensiveLayer);
      expect(minimalLayer.layerVersionArn).not.toBe(
        comprehensiveLayer.layerVersionArn,
      );
    });
  });
});
