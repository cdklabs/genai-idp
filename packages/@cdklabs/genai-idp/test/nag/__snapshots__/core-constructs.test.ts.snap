// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Core Constructs CDK Nag Compliance IdpPythonLayerVersion satisfies CDK NAG AwsSolutionsChecks with minimal configuration 1`] = `
{
  "Parameters": {
    "BootstrapVersion": {
      "Default": "/cdk-bootstrap/hnb659fds/version",
      "Description": "Version of the CDK Bootstrap resources in this environment, automatically retrieved from SSM Parameter Store. [cdk:skip]",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
  },
  "Resources": {
    "comamazonawscdkcdklabsgenaiidpidpcommonlayer1695a11fC9CF6570": {
      "Metadata": {
        "cdk_nag": {
          "rules_to_suppress": [
            {
              "id": "AwsSolutions-L1",
              "reason": "Lambda layer uses supported Python runtime version as required by the IDP framework",
            },
          ],
        },
      },
      "Properties": {
        "CompatibleRuntimes": [
          "python3.12",
        ],
        "Content": {
          "S3Bucket": {
            "Fn::Sub": "cdk-hnb659fds-assets-\${AWS::AccountId}-\${AWS::Region}",
          },
          "S3Key": "30b3ba90159d31e284ee43e6f518f32b456050143141edcb11f0d59b811e6135.zip",
        },
        "Description": "Lambda Layer containing the idp_common Python package with modules: core",
      },
      "Type": "AWS::Lambda::LayerVersion",
    },
  },
  "Rules": {
    "CheckBootstrapVersion": {
      "Assertions": [
        {
          "Assert": {
            "Fn::Not": [
              {
                "Fn::Contains": [
                  [
                    "1",
                    "2",
                    "3",
                    "4",
                    "5",
                  ],
                  {
                    "Ref": "BootstrapVersion",
                  },
                ],
              },
            ],
          },
          "AssertDescription": "CDK bootstrap stack version 6 required. Please run 'cdk bootstrap' with a recent version of the CDK CLI.",
        },
      ],
    },
  },
}
`;

exports[`Core Constructs CDK Nag Compliance IdpPythonLayerVersion satisfies CDK NAG with comprehensive module configuration 1`] = `
{
  "Parameters": {
    "BootstrapVersion": {
      "Default": "/cdk-bootstrap/hnb659fds/version",
      "Description": "Version of the CDK Bootstrap resources in this environment, automatically retrieved from SSM Parameter Store. [cdk:skip]",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
  },
  "Resources": {
    "comamazonawscdkcdklabsgenaiidpidpcommonlayeref7b2694B946B9A8": {
      "Metadata": {
        "cdk_nag": {
          "rules_to_suppress": [
            {
              "id": "AwsSolutions-L1",
              "reason": "Lambda layer uses supported Python runtime version as required by the IDP framework with multiple modules",
            },
          ],
        },
      },
      "Properties": {
        "CompatibleRuntimes": [
          "python3.12",
        ],
        "Content": {
          "S3Bucket": {
            "Fn::Sub": "cdk-hnb659fds-assets-\${AWS::AccountId}-\${AWS::Region}",
          },
          "S3Key": "30b3ba90159d31e284ee43e6f518f32b456050143141edcb11f0d59b811e6135.zip",
        },
        "Description": "Lambda Layer containing the idp_common Python package with modules: assessment, classification, ocr",
      },
      "Type": "AWS::Lambda::LayerVersion",
    },
  },
  "Rules": {
    "CheckBootstrapVersion": {
      "Assertions": [
        {
          "Assert": {
            "Fn::Not": [
              {
                "Fn::Contains": [
                  [
                    "1",
                    "2",
                    "3",
                    "4",
                    "5",
                  ],
                  {
                    "Ref": "BootstrapVersion",
                  },
                ],
              },
            ],
          },
          "AssertDescription": "CDK bootstrap stack version 6 required. Please run 'cdk bootstrap' with a recent version of the CDK CLI.",
        },
      ],
    },
  },
}
`;
